### PR TITLE
fix(es/resolver): Resolve shadowing of `undefined`, `NaN`, and `Infinity`

### LINKED ISSUE #8465: wrongly transform function Infinity to 1/0
### Describe the bug
Infinity is wrongly transformed to 1/0
### Input code

```typescript
function Infinity(){
    console.log('xxx');
}
export default Infinity;
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript",
      "jsx": false
    },
    "target": "es5",
    "loose": false,
    "minify": {
      "compress": {
        "arguments": false,
        "arrows": true,
        "booleans": true,
        "booleans_as_integers": false,
        "collapse_vars": true,
        "comparisons": true,
        "computed_props": true,
        "conditionals": true,
        "dead_code": true,
        "directives": true,
        "drop_console": false,
        "drop_debugger": true,
        "evaluate": true,
        "expression": false,
        "hoist_funs": false,
        "hoist_props": true,
        "hoist_vars": false,
        "if_return": true,
        "join_vars": true,
        "keep_classnames": false,
        "keep_fargs": true,
        "keep_fnames": false,
        "keep_infinity": false,
        "loops": true,
        "negate_iife": true,
        "properties": true,
        "reduce_funcs": false,
        "reduce_vars": false,
        "side_effects": true,
        "switches": true,
        "typeofs": true,
        "unsafe": false,
        "unsafe_arrows": false,
        "unsafe_comps": false,
        "unsafe_Function": false,
        "unsafe_math": false,
        "unsafe_symbols": false,
        "unsafe_methods": false,
        "unsafe_proto": false,
        "unsafe_regexp": false,
        "unsafe_undefined": false,
        "unused": true,
        "const_to_let": true,
        "pristine_globals": true
      },
      "mangle": false
    }
  },
  "module": {
    "type": "es6"
  },
  "minify": false,
  "isModule": true
}
```

### Playground link (or link to the minimal reproduction)

https://play.swc.rs/?version=1.3.100&code=H4sIAAAAAAAAA0srzUsuyczPU%2FDMS8vMyyyp1NCs5lIAguT8vOL8nFS9nPx0DfWKigp1TWuuWq7UioL8ohKFlNS0xNKcErgmawCsssEKRwAAAA%3D%3D&config=H4sIAAAAAAAAA32US5LjIAyG932KlNe9nVnMAXo3Z6AICIc0Ri4k0nF15e4jY%2BcxHTk7W59%2BCSSh77fdrjuS6%2F7svuVTfkZbCMrtXyw0ZbZnsXTgBkuuxJG79ys90oyCTQTNdFlIx7b0wE1Fv1b3LiESXN1X2xBzDNNjQofDWIDowSZWiVcHyEz%2F61dW8GsGXOqjfY%2BYwOYXxFgyMTP0ULTADlOyI4E52aJEmU9qSyTUUsywMngzFhxVnn3kiFlyPlMP1huHHhQUCziOJ9BkkktkmeR6yn0a9rCvfd%2Ba%2FEMNJ5uqZSUnnFtL5LRK1ANGYhNq1kq4wI0aLHAt7k9lDKYA15KfdUeMeaMnnwBSgWSJsh1Ai9s8gszTljq8VMYcZGR5UrjMt3bLDL0U1cQYlMrOlYHCUetmAV8dzJV12nFWvFE%2Bih4MhCCzooSmr8juoCXlaQQMCpD%2B2qBN1QLM7RVu8PlBvMAfckvWB2z1GCwftilNwx7TiwQD8AH9CwdpBeM2LrIlzuM2r9mDjAZ41aVSA89LQB4Ao0ltWT7NhjwPiWj6hPv7mlgdLrclPNjc39%2F7soffVoduQF8bXDf83N9lL%2F%2Fu7k7XLXw7eBfp71XZkl7%2BAVPL6HotBgAA

### SWC Info output

1.3.100

### Expected behavior

don't turn to 1/0 like terser 
```js
function o(){console.log("xxx")}export default o;
```

### Actual behavior
```js
export default 1 / 0;
```

### Version

1.3.100

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by magic-akari**
**Related issue:**
- Closes: #8465

### CODE REVIEW

**Comment by magic-akari** [comment]
Investigation:

For cases that are not top-level, SWC is already capable of handling them effectively. It addresses the issue related to shadowing.

[Playground](https://play.swc.rs/?version=1.3.100&code=H4sIAAAAAAAAA6vmUgCCssQiBc%2B8tMy8zJJKBVsFQ2uwaHJ%2BXnF%2BTqpeTn66BkxW01pBXx%2BulquWqxqsNCe1hCQDDLlqAYa7ySl6AAAA&config=H4sIAAAAAAAAA32US5LjIAyG932KlNe9nVnMAXo3Z6AICIc0Ri4k0nF15e4jY%2BcxHTk7W59%2BCSSh77fdrjuS6%2F7svuVTfkZbCMrtXyw0ZbZnsXTgBkuuxJG79ys90oyCTQTNdFlIx7b0wE1Fv1b3LiESXN1X2xBzDNNjQofDWIDowSZWiVcHyEz%2F61dW8GsGXOqjfY%2BYwOYXxFgyMTP0ULTADlOyI4E52aJEmU9qSyTUUsywMngzFhxVnn3kiFlyPlMP1huHHhQUCziOJ9BkkktkmeR6yn0a9rCvfd%2Ba%2FEMNJ5uqZSUnnFtL5LRK1ANGYhNq1kq4wI0aLHAt7k9lDKYA15KfdUeMeaMnnwBSgWSJsh1Ai9s8gszTljq8VMYcZGR5UrjMt3bLDL0U1cQYlMrOlYHCUetmAV8dzJV12nFWvFE%2Bih4MhCCzooSmr8juoCXlaQQMCpD%2B2qBN1QLM7RVu8PlBvMAfckvWB2z1GCwftilNwx7TiwQD8AH9CwdpBeM2LrIlzuM2r9mDjAZ41aVSA89LQB4Ao0ltWT7NhjwPiWj6hPv7mlgdLrclPNjc39%2F7soffVoduQF8bXDf83N9lL%2F%2Fu7k7XLXw7eBfp71XZkl7%2BAVPL6HotBgAA)

```JavaScript
{
    var Infinity = 1;
    console.log(Infinity); // Infinity
}
{
    let Infinity = 1;
    console.log(Infinity); // 1
}
```

For the top-level cases, it is necessary to consider different scenarios. 
When it comes to browsers, the result can be affected by whether it is an ESM or not.

```html
<script>
console.log(Infinity); //  SyntaxError: Identifier 'Infinity' has already been declared
function Infinity() {}
</script>
```

```html
<script type="module">
console.log(Infinity); // ƒ Infinity() { }
function Infinity() { }
</script>
```

For Node.js, whether it's ESM or not, `Infinity`` is always a shadowed reference.

```JavaScript
console.log(Infinity); // [Function: Infinity]
function Infinity() { }
```

**Comment by magic-akari** [comment]
A preferable approach is to never rename top-level `undefined`, `NaN`, and `Infinity`.

File: crates/swc_ecma_minifier/src/lib.rs [hunk 1]
```diff
@@ -233,7 +233,13 @@ pub fn optimize(
...
         // TODO: base54.reset();
 
-        let preserved = idents_to_preserve(mangle.clone(), &n);
+        let mut preserved = idents_to_preserve(mangle.clone(), &n);
+
+        // https://tc39.es/ecma262/multipage/global-object.html#sec-value-properties-of-the-global-object-infinity
+        let unresolved_ctxt = SyntaxContext::empty().apply_mark(marks.unresolved_mark);
+        preserved.insert(("undefined".into(), unresolved_ctxt));
+        preserved.insert(("NaN".into(), unresolved_ctxt));
+        preserved.insert(("Infinity".into(), unresolved_ctxt));
```
Comment (Austaras) [inline crates/swc_ecma_minifier/src/lib.rs:242]: Why need this here?

Comment (magic-akari) [reply crates/swc_ecma_minifier/src/lib.rs:242]: SWC will rename the variable even it's unresolved.
Cannot find the reason now.

File: crates/swc_ecma_minifier/tests/fixture/issues/8465/output.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+function Infinity() {
+    console.log("xxx");
+}
+export default Infinity;
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/fixture/issues/8465/output.js:4]: This doesn't seem like to be the optimal output. It should be
```js
export default function() {
    console.log("xxx");
};
```

Comment (magic-akari) [reply crates/swc_ecma_minifier/tests/fixture/issues/8465/output.js:4]: check this https://github.com/swc-project/swc/pull/8469/commits/1aefb0a7ccca8ea15ebb135c28ce88a8973adbf9#diff-0c587eaeb38866a706c0a8355bf41885ef6871afd928e60a5975fe1a46eb0d1fR1-R3

File: crates/swc_ecma_minifier/src/lib.rs (hunk 1 above)
Comment (kdy1) [reply crates/swc_ecma_minifier/src/lib.rs:242]: Can we move this list to https://github.com/swc-project/swc/blob/1aefb0a7ccca8ea15ebb135c28ce88a8973adbf9/crates/swc_ecma_transforms_base/src/rename/analyzer/scope.rs#L201-L216 ?

Comment (magic-akari) [reply crates/swc_ecma_minifier/src/lib.rs:242]: I just attempted it. Passing the `unresolved_mark` deeply to this function requires modifications at multiple layers.

Comment (kdy1) [reply crates/swc_ecma_minifier/src/lib.rs:242]: Ah, I see

File: crates/swc_ecma_minifier/tests/terser/compress/issue_597/beautify_off_1/output.js [hunk 3]
```diff
@@ -1,2 +1,2 @@
-var NaN1;
+var NaN;
 console.log(null, void 0, 1 / 0, NaN, NaN, (1 / 0).toString(), NaN.toString(), NaN.toString());
```
Comment (Austaras) [inline crates/swc_ecma_minifier/tests/terser/compress/issue_597/beautify_off_1/output.js:2]: The input here is
```js
console.log(null, undefined, Infinity, NaN, Infinity * undefined, ...)
)
```
The 4th and 5th arg should be different.

Comment (magic-akari) [reply crates/swc_ecma_minifier/tests/terser/compress/issue_597/beautify_off_1/output.js:2]: crates/swc_ecma_minifier/tests/terser/compress/issue_597/beautify_off_1/output.terser.js

This is another issue. When you examine the output of Terser, you will discover that Terser avoids pointing to the global NaN when it encounters user-defined NaN.

```JavaScript
var NaN;
console.log(
    null,
    void 0,
    1 / 0,
    NaN,
    0 / 0,
    (1 / 0).toString(),
    NaN.toString(),
    (0 / 0).toString()
);
```

crates/swc_ecma_minifier/tests/terser/compress/issue_597/NaN_and_Infinity_should_not_be_replaced_when_they_are_redefined_evaluate

The intention can be inferred from the name of this test.

Comment (Austaras) [reply crates/swc_ecma_minifier/tests/terser/compress/issue_597/beautify_off_1/output.js:2]: That's in script mode.

**Comment by magic-akari** [comment]
close for preferring #8471

**Pull request closed without merging**